packages/@expo/cli/src/start/server/type-generation/routes.ts

```diff
@@ -0,0 +1,322 @@
+import fs from 'fs/promises';
+import { debounce } from 'lodash';
+import { Server } from 'metro';
+import path from 'path';
+
+import { env } from '../../../utils/env';
+import { unsafeTemplate } from '../../../utils/template';
+import { ServerLike } from '../BundlerDevServer';
+import { metroWatchTypeScriptFiles } from '../metro/metroWatchTypeScriptFiles';
+import { setToUnionType } from './utils';
+
+// /test/[...param1]/[param2]/[param3] - captures ["param1", "param2", "param3"]
+export const CAPTURE_DYNAMIC_PARAMS = /\[(?:\.{3})?(\w*?)[\]$]/g;
+// /[...param1]/ - Match [...param1]
+export const CATCH_ALL = /\[\.\.\..+?\]/g;
+// /[param1] - Match [param1]
+export const SLUG = /\[.+?\]/g;
+// /(group1,group2,group3)/test - match (group1,group2,group3)
+export const ARRAY_GROUP_REGEX = /\(\w+?,.*?\)/g;
+// /(group1,group2,group3)/test - captures ["group1", "group2", "group3"]
+export const CAPTURE_GROUP_REGEX = /[\\(,](\w+?)(?=[,\\)])/g;
+
+export interface SetupTypedRoutesOptions {
+  server: ServerLike;
+  metro: Server;
+  typesDirectory: string;
+}
+
+export async function setupTypedRoutes({ server, metro, typesDirectory }: SetupTypedRoutesOptions) {
+  const appRoot = path.resolve(env.EXPO_ROUTER_APP_ROOT);
```

Comment:
This value is relative to the `node_module/expo-router/entry` file, so it cannot be resolved like this. The actual resolution is more like this https://github.com/expo/expo/blob/036e9444308bd944b04f0d4537f0c055e344b8d0/packages/%40expo/cli/src/start/server/metro/router.ts#LL18C20-L18C51

Can add a function to that file like `getExpoRouterRootDirectory(projectRoot: string): string` which just returns `path.join(projectRoot, 'app')` then in the future we can extend it to support custom directories.